Add a tool to run `x.py` from any subdirectory

casey opened the pull request:
This adds a binary called `x` in `src/tools/x`. All it does is check the current directory and its ancestors for a file called `x.py`, and if it finds one, runs it.

By installing x, you can easily run `x.py` from any subdirectory, and only need to type `x`.

It can be installed with `cargo install --path src/tools/x`

This is a copy of a [binary I've been using myself when working on rust](https://github.com/casey/bootstrap), currently published to crates.io as `bootstrap`.

It could be changed to avoid indirecting through `x.py`, and instead call the bootstrap module directly. However, this seemed like the simplest thing possible, and won't break if the details of how the bootstrap module is invoked change.

rust-highfive commented:
r? @Mark-Simulacrum

(rust_highfive has picked a reviewer for you, use r? to override)

jyn514 commented:
Assuming this is accepted, can you also make a PR to the rustc-dev-guide documenting how to use it? It would go in https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html.

casey commented:
> Assuming this is accepted, can you also make a PR to the rustc-dev-guide documenting how to use it? It would go in https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html.

Yup, definitely, I'd be happy to.

jyn514 commented on src/tools/x/src/main.rs:
```diff
@@ -0,0 +1,32 @@
+    };
+
+    for dir in current.ancestors() {
+        let candidate = dir.join("x.py");
+
+        if candidate.exists() {
+            let error = Command::new(&candidate).args(env::args().skip(1)).current_dir(dir).exec();
```
While you're being clever, `./x.py` doesn't work on Windows - can you change this to run `python x.py` directly?

jyn514 replied:
And actually, if you could look up `python`, `python3`, `python2` and look up the first one that works that would be _amazing_. https://github.com/rust-lang/rust/issues/71818

casey replied:
Done!

jyn514 commented on src/tools/x/Cargo.lock:
```diff
@@ -0,0 +1,5 @@
+# This file is automatically @generated by Cargo.
+# It is not intended for manual editing.
+[[package]]
+name = "bootstrap"
+version = "1.0.2"
```
This is left over from experimenting I think, `Cargo.toml` has no such dependency.

casey replied:
Fixed

jyn514 commented on src/tools/x/src/main.rs:
```diff
@@ -1,9 +1,27 @@
+    process::{self, Command, ExitStatus},
 };
 
+trait CommandExt {
+    fn exec_or_status(&mut self) -> io::Result<ExitStatus>;
+}
+
+#[cfg(unix)]
+impl CommandExt for Command {
+    fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
+        use std::os::unix::process::CommandExt;
+        Err(self.exec())
+    }
+}
+
+#[cfg(not(unix))]
+impl CommandExt for Command {
+    fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
+        self.status()
+    }
+}
```
This doesn't need to be a trait I think, you can just have two free functions.

```suggestion
#[cfg(unix)]
fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
    use std::os::unix::process::CommandExt;
    Err(self.exec())
}

#[cfg(not(unix))]
fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
    self.status()
}
```

casey replied:
fixed

casey commented:
Thanks for the review! I think I addressed everything.

Mark-Simulacrum commented on src/tools/x/src/main.rs:
```diff
@@ -0,0 +1,92 @@
+#[cfg(not(unix))]
+fn exec_or_status(command: &mut Command) -> io::Result<ExitStatus> {
+    command.status()
+}
```
Can we just always invoke status() here? I think the extra process shouldn't matter at all, right?

casey replied:
We could invoke status, but if users are using something like `ps | grep` to see running processes, extra processes can get annoying, and make it confusing to know what to kill.

Mark-Simulacrum replied:
Ok, resolving this -- I think this seems fine. We do this inside bootstrap I think already and it's not exactly lots of code either :)

Mark-Simulacrum reviewed (commented):
I am a bit worried this might break the process group creation on Windows where we attempt to associate rustbuild with the Python process (IIRC) -- can you check if ctrl+c on Windows when running through this tool kills Cargo and any running rustc processes?

Mark-Simulacrum commented:
I am generally okay with adding this but if it's a cargo project you won't be able to run it (in general) without running x.py first unless you're manually managing submodules (otherwise Cargo.toml will not resolve). I am a bit hesitant as a result to add it. But not really opposed, it seems low maintenance cost, modulo my comments on Windows process scheduling.

jyn514 commented:
> if it's a cargo project you won't be able to run it (in general) without running x.py first unless you're manually managing submodules (otherwise Cargo.toml will not resolve)

Could we *not* make it part of the workspace, so it doesn't need to resolve any project other than its own directory?

Mark-Simulacrum commented:
That would work, I think, yes. I suppose the downside is that we can't make it the default-run in the workspace, but since that's not a good idea anyway due to the submodule problem, seems fine.

Please also squash commits.

jyn514 commented:
We could always add it to the workspace if we figure out the submodule problem later.

camelid commented:
> I think the name `x`, is not a common binary name. There is `X`, I think usually a shortcut that launches X11, but on a case-sensitive system, there isn't any conflict.

I want to note that macOS is unfortunately case-insensitive, so this doesn’t hold for macOS.

jyn514 commented:
@camelid X11 isn't installed on MacOS, though, right?

camelid commented:
I don't think it's installed by default, but people may install it themselves (see also [XQuartz](https://www.xquartz.org/)). I just wanted to note that macOS is case-insensitive, so this will collide with anything named `x` or `X`/

jyn514 commented:
I think that's a reasonable tradeoff - you'd have to install _both_ of the binaries yourself, at which point you'd know that they overlap.

jyn514 commented:
@casey I think this is waiting on you to remove src/tools/x from the workspace and test that it can be run without first running x.py.

casey commented:
I checked on windows, and `ctrl-c` seems to work when running the wrapper.

I removed it from the workspace, and added it to the exclude list. It now works when run in a fresh checkout, before submodules are initialized.

I also squashed the commits.

Mark-Simulacrum commented:
@bors r+ rollup

Thanks!

bors commented:
:pushpin: Commit 5fc22f1431e209610040d92ba671966ab2781f0b has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 5fc22f1431e209610040d92ba671966ab2781f0b -->
<!-- homu: {"type":"Approved","sha":"5fc22f1431e209610040d92ba671966ab2781f0b","approver":"Mark-Simulacrum"} -->

casey commented:
Nice! Thank you both for reviewing ^_^

jyn514 commented:
@casey do you mind making that PR to the dev-guide?

casey commented:
@jyn514 Yup, definitely, it's on my todo list.

casey commented:
I just opened https://github.com/rust-lang/rustc-dev-guide/pull/945 to document this.
